Language/Haskell/GhcMod/ImportedFrom.hs

```diff
@@ -0,0 +1,785 @@
+  where
+    leadingDot :: String
+    leadingDot = '.' : last (separateBy '.' symb)
+
+    -- f symbol export = filter (symbol ==) thisExports
+    f symbol export = filter (symbol `isSuffixOf`) thisExports
+       where thisExports = qualifiedExports export         -- Things that this module exports.
+
+refineVisibleExports
+    :: forall m. (GhcMonad m, MonadIO m, GmOut m, GmLog m)
+    => (String -> m (Maybe String))
+    -> [ModuleExports]
+    -> m [ModuleExports]
+refineVisibleExports getHaddockInterfaces exports = mapM f exports
+  where
+    f :: ModuleExports -> m ModuleExports
+    f mexports = do
+        let pname          = mPackageName     mexports -- e.g. "base-4.8.2.0"
+            thisModuleName = mName            mexports -- e.g. "Prelude"
+            qexports       = qualifiedExports mexports -- e.g. ["base-4.8.2.0:GHC.Base.Just", ...]
+        visibleExportsMap <- getVisibleExports getHaddockInterfaces pname
+        gmLog GmDebug "" $ strDoc $ "visibleExportsMap: " ++ show visibleExportsMap
+
+        let thisModVisibleExports = fromMaybe
+                                        (error $ "Could not get visible exports of " ++ pname)
+                                        (join $ traverse (M.lookup thisModuleName) visibleExportsMap)
+
+        let qexports' = filter (hasPostfixMatch thisModVisibleExports) qexports
+
+        gmLog GmDebug "" $ strDoc $ show (qexports, qexports')
+
+        return $ mexports { qualifiedExports = qexports' }
+
+    -- hasPostfixMatch "base-4.8.2.0:GHC.Base.Just" ["Just", "True", ...] -> True
+    hasPostfixMatch :: [String] -> String -> Bool
+    hasPostfixMatch xs s = last (separateBy '.' s) `elem` xs
+
+-- | The last thing with a single export must be the match? Iffy.
+getLastMatch :: [ModuleExports] -> Maybe ModuleExports
+getLastMatch exports = Safe.lastMay $ filter f exports
+  where
+    f me = length (qualifiedExports me) == 1
+
+-- | Try to look up the Haddock URL for a symbol.
+guessHaddockUrl
+    :: forall m.
+       (GhcMonad m, MonadIO m, GmOut m, GmLog m)
+    => ModSummary
+    -> FilePath
+    -> String
+    -> String
+    -> Int
+    -> Int
+    -> FilePath
+    -> (FilePath -> [String] -> String -> IO String)
+    -> [GhcPkgDb]
+    -> m (Either String String)
+guessHaddockUrl modSum targetFile targetModule symbol lineNr colNr ghcPkg readProc pkgDbStack = do
+    gmLog GmDebug "" $ strDoc $ "guessHaddockUrl: targetFile: "   ++ targetFile
+    gmLog GmDebug "" $ strDoc $ "guessHaddockUrl: targetModule: " ++ targetModule
+    gmLog GmDebug "" $ strDoc $ "guessHaddockUrl: symbol: "       ++ show symbol
+    gmLog GmDebug "" $ strDoc $ "guessHaddockUrl: line nr: "      ++ show lineNr
+    gmLog GmDebug "" $ strDoc $ "guessHaddockUrl: col nr: "       ++ show colNr
+
+    let textualImports  = ms_textual_imps modSum
+        importDecls0 = map toImportDecl textualImports
```

Comment:
You have a `GhcMonad m` here. Use it for proper pretty printing instead of using unsafeGlobalDynFlags please. Unless I'm missing a reason everything should always be qualified or something?